fusillade/clouddirectory.py

```diff
@@ -0,0 +1,807 @@
+        """
+
+        :param object_ref: The reference that identifies the object.
+        :param update_params:
+        :return:
+        """
+        updates = []
+        for i in update_params:
+            updates.append(
+                {
+                    'ObjectAttributeKey': {
+                        'SchemaArn': self._schema,
+                        'FacetName': i.facet,
+                        'Name': i.attribute
+                    },
+                    'ObjectAttributeAction': {
+                        'ObjectAttributeActionType': i.action.name,
+                        'ObjectAttributeUpdateValue': {
+                            i.value_type.name: i.value
+                        }
+                    }
+                }
+            )
+        return ad.update_object_attributes(
+            DirectoryArn=self._dir_arn,
+            ObjectReference={
+                'Selector': object_ref
+            },
+            AttributeUpdates=updates
+        )
+
+    def create_folder(self, path: str, name: str) -> None:
+        # A folder is just a Group
+        schema_facets = [dict(SchemaArn=self._schema, FacetName="Group")]
+        object_attribute_list = self._get_object_attribute_list(facet="Group", name=name)
+        try:
+            ad.create_object(DirectoryArn=self._dir_arn,
+                             SchemaFacets=schema_facets,
+                             ObjectAttributeList=object_attribute_list,
+                             ParentReference=dict(Selector=path),
+                             LinkName=name)
+        except ad.exceptions.LinkNameAlreadyInUseException:
+            pass
+
+    def clear(self) -> None:
+        for policy_ref in self.list_object_children('/Policies/'):
+            self.delete_policy(policy_ref)
+        for obj_ref in self.list_object_children('/Users/'):
+            self.delete_object(obj_ref)
+        for obj_ref in self.list_object_children('/Groups/'):
+            self.delete_object(obj_ref)
+
+    def delete_policy(self, policy_ref: str) -> None:
+        self.batch_write([self.batch_detach_policy(policy_ref, obj_ref)
+                          for obj_ref in self.list_policy_attachments(policy_ref)])
+        self.batch_write([self.batch_detach_object(parent_ref, link_name)
+                          for parent_ref, link_name in self.list_object_parents(policy_ref)])
+        ad.delete_object(DirectoryArn=self._dir_arn, ObjectReference={'Selector': policy_ref})
+
+    def delete_object(self, obj_ref: str) -> None:
+        self.batch_write([self.batch_detach_policy(policy_ref, obj_ref)
+                          for policy_ref in self.list_object_policies(obj_ref)])
+        self.batch_write([self.batch_detach_object(parent_ref, link_name)
+                          for parent_ref, link_name in self.list_object_parents(obj_ref)])
+        ad.delete_object(DirectoryArn=self._dir_arn, ObjectReference={'Selector': obj_ref})
```

Comment:
What cases are the policy and object detachment addressing that are not addressed inside of `delete_object`?  Maybe just a comment here explaining.

Reply:
there are a bunch of requirements before something can be deleted.
1. removed from tree
1. no links
1. no children
1. no attributes...

for more detail see "Deleting Object" [here](https://docs.aws.amazon.com/clouddirectory/latest/developerguide/directory_objects_access_objects.html)

Reply:
delete_object uses list_object_policies to find policies attached to it
delete_policy  uses list_policy_attachments to find what objects it's attached to. 

Linking object in cloud directory is directional

Reply:
I think it might be good to include as a comment.

Reply:
adding doc